File: 402-remove-k-digits/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        StringBuilder sbd = new StringBuilder();
        for(char ch: num.toCharArray()){
            while(k>0 && sbd.length()>0 && sbd.charAt(sbd.length()-1)>ch){
                sbd.deleteCharAt(sbd.length()-1);
                k--;
            }
            sbd.append(ch);
        }

        while(k>0 && sbd.length()>0){
            sbd.deleteCharAt(sbd.length()-1);
            k--;
        }

        int firstNonZeroIndex = 0;
        while(firstNonZeroIndex<sbd.length() && sbd.charAt(firstNonZeroIndex)=='0'){
            firstNonZeroIndex++;
        }

        sbd.delete(0,firstNonZeroIndex);

        if(sbd.length()==0){
            sbd = new StringBuilder("0");
        }

        return sbd.toString();
    }
}

File: 402-remove-k-digits/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        string ans;
        for(char& ch: num){
            while(k>0 && ans.size()>0 && ans.back()>ch){
                ans.pop_back();
                k--;
            }

            ans += ch;
        }

        if(!ans.empty() && ans[0]=='0'){
            int firstNonZeroIndex = 0;
            while(firstNonZeroIndex<n && ans[firstNonZeroIndex]=='0'){
                firstNonZeroIndex++;
            }

            ans = ans.substr(firstNonZeroIndex);
        }

        while(k>0 && !ans.empty()){
            ans.pop_back();
            k--;
        }

        if(ans.empty()){
            ans += '0';
        }

        return ans;
    }
};
